zkevm-circuits/src/evm_circuit/op_execution.rs

```diff
@@ -0,0 +1,532 @@
+    pub program_counter: Cell<F>,
+    pub stack_pointer: Cell<F>,
+    gas_counter: Cell<F>,
+    opcode: Cell<F>,
+}
+
+impl<F: FieldExt> OpExecutionState<F> {
+    pub(crate) fn new(cells: &[Cell<F>]) -> Self {
+        assert_eq!(cells.len(), NUM_CELL_OP_EXECTION_STATE);
+
+        Self {
+            is_executing: cells[0].clone(),
+            global_counter: cells[1].clone(),
+            call_id: cells[2].clone(),
+            program_counter: cells[3].clone(),
+            stack_pointer: cells[4].clone(),
+            gas_counter: cells[5].clone(),
+            opcode: cells[6].clone(),
+        }
+    }
+}
+
+trait OpGadget<F: FieldExt> {
+    const RESPONSIBLE_OPCODES: &'static [u8];
+
+    const CASE_CONFIGS: &'static [CaseConfig];
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self;
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>>;
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        core_state: &mut CoreStateInstance,
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error>;
+}
+
+// Preset stores default values for each case of op gadget
+#[derive(Clone, Default)]
+struct Preset<F> {
+    qs_byte_lookups: [F; CIRCUIT_HEIGHT],
+    free_cells: Vec<(usize, F)>,
+}
+
+#[derive(Clone)]
+pub(crate) struct OpExecutionGadget<F> {
+    r: F,
+    qs_byte_lookups: Vec<Cell<F>>,
+    state_curr: OpExecutionState<F>,
+    state_next: OpExecutionState<F>,
+    qs_ops: Vec<Cell<F>>,
+    free_cells: Vec<Cell<F>>,
+    resumption: Resumption<F>,
+    // TODO: use OpcodeId from bus_mapping
+    qs_op_idx_map: HashMap<u8, usize>,
+    preset_map: HashMap<(usize, Case), Preset<F>>,
+    add_gadget: AddGadget<F>,
+    push_gadget: PushGadget<F>,
+}
+
+impl<F: FieldExt> OpExecutionGadget<F> {
+    // TODO: refactor input type
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn configure(
+        meta: &mut ConstraintSystem<F>,
+        r: F,
+        qs_op_execution: Expression<F>,
+        qs_byte_lookups: Vec<Cell<F>>,
+        state_curr: OpExecutionState<F>,
+        state_next: OpExecutionState<F>,
+        free_cells: Vec<Cell<F>>,
+        independent_lookups: &mut Vec<(Expression<F>, Vec<Lookup<F>>)>,
+    ) -> Self {
+        let (qs_ops, free_cells) =
+            free_cells.split_at(NUM_CELL_OP_GADGET_SELECTOR);
+        let resumption = Resumption::new(
+            &free_cells[free_cells.len() - NUM_CELL_RESUMPTION..],
+        );
+
+        let mut qs_op_idx_map = HashMap::new();
+        let mut preset_map = HashMap::new();
+        let mut qs_op_idx = 0;
+
+        let mut constraints = vec![Constraint {
+            name: "op selectors",
+            selector: Expression::Constant(F::one()),
+            polys: bool_switches_constraints(qs_ops),
+            lookups: vec![],
+        }];
+
+        macro_rules! construct_op_gadget {
```

Comment:
We should document macros as much as possible.

Reply:
Added in [`29b25cd`](https://github.com/appliedzkp/zkevm-circuits/pull/26/commits/29b25cd033b82b7a4e2187418a5ae6d5715253ae)